File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int answer = 100;

vector<string> get_next_words(string word, vector<string> words){
    vector<string> next_word_list;
    for (string w: words){
        int count = 0;
        for (int i = 0; i < word.size(); i++){
            if (word[i] != w[i]){
                count ++;
            }
        }
        if (count == 1){
            next_word_list.push_back(w);
        }
    }
    return next_word_list;
}

vector<string> get_subvector(string word, vector<string> words){
    words.erase(find(words.begin(), words.end(), word));
    return words;
}

void dfs(string curr, string target, vector<string> words, int count){
    if (curr == target){
        if (answer > count){
            answer = count;
        }
    }
    
    vector<string> next_word_list = get_next_words(curr, words);
    
    for (string next: next_word_list){
        dfs(next, target, get_subvector(next, words), count + 1);
    }
}

int solution(string begin, string target, vector<string> words) {
    if (find(words.begin(), words.end(), target) == words.end()){
        return 0;
    }
    dfs(begin, target, words, 0);
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        // target이 words에 없으면 0을 리턴
        if (!Arrays.asList(words).contains(target)) {
            return 0;
        }

        // dfs로 탐색
        List<String> wordList = new ArrayList<>(Arrays.asList(words));
        wordList.remove(begin);
        int result = search(begin, target, wordList, 0);
        if (result > 50) return 0;
        else return result;
    }

    public int search(String curr, String target, List<String> words, int count) {
        if (words.size() == 0) {
            return 100;
        }

        if (curr.equals(target)) {
            return count;
        }

        int result = 100;
        for (String word : words) {
            if (getCountOfDiffChars(curr, word) == 1) {
                List<String> newWords = new ArrayList<>(words);
                newWords.remove(word);
                result = Math.min(result, search(word, target, newWords, count + 1));
            }
        }
        return result;
    }

    public int getCountOfDiffChars(String a, String b) {
        int count = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                count++;
            }
        }
        return count;
    }
}